Initial support for gen 2 products

w1ll1am23 opened the pull request:
This is in no way even close to ready to merge but I am getting the devices in HA without errors. Very little works yet.

Using pipes for the types isn't working on my system for some reason... so currently just setting things to RestIot to get it working.


Associated library code.
https://github.com/dahlb/hatch_rest_api/pull/6

dahlb commented:
don't have a gen 2 so I can't help much, but it looks like your local issue with pipes (union) is probably that you are programming an earlier version of python then 3.10 https://peps.python.org/pep-0604/

w1ll1am23 commented:
@dahlb Ahh that makes sense the library shows it works with 3.9 but this integration doesn't so this makes sense. I'll have to upgrade to 3.10 I was having issues setting up a virtualenv with 3.10 I'll get that figured out.

Hopefully I manage to get this all working once I think things are ready I'll remove the draft label and tag you.

trevorglaser commented:
@w1ll1am23 I am getting this error when trying to test your new code.

`Unable to prepare setup for platform ha_hatch.media_player: Platform not found (cannot import name 'RestIot' from 'hatch_rest_api' (/srv/homeassistant/lib/python3.9/site-packages/hatch_rest_api/__init__.py)).
Unable to prepare setup for platform ha_hatch.light: Platform not found (cannot import name 'RestIot' from 'hatch_rest_api' (/srv/homeassistant/lib/python3.9/site-packages/hatch_rest_api/__init__.py)).
Unable to prepare setup for platform ha_hatch.binary_sensor: Platform not found (cannot import name 'RestIot' from 'hatch_rest_api' (/srv/homeassistant/lib/python3.9/site-packages/hatch_rest_api/__init__.py)).
Unable to prepare setup for platform ha_hatch.sensor: Platform not found (cannot import name 'RestIot' from 'hatch_rest_api' (/srv/homeassistant/lib/python3.9/site-packages/hatch_rest_api/__init__.py)).
Unable to prepare setup for platform ha_hatch.switch: Platform not found (cannot import name 'RestIot' from 'hatch_rest_api' (/srv/homeassistant/lib/python3.9/site-packages/hatch_rest_api/__init__.py)).`

w1ll1am23 commented:
Sorry I'll try to get back to this shortly. Ran in to several road blocks. Firstly I bricked my unit and had to work with support to get it fixed.

After that I did some digging trying to get full support for all Sleep sounds not just favorites and that resulted in me bricking my unit again... 

I think getting full support for all sounds is going to be impossible, it looks like Hatch is using an additional third party API called contentful and the URLS for those sounds are generated by some call to that third party. Pulling that in here would be a lot of extra work. I think limiting this integration to controlling the light and the predefined "favorites" is going to be the best bet. That will result in non-favorite sounds set via the app showing HA in a playing unknown state. It will also make any interaction with the media player make changes to the light entity as well since a favorite has settings for all of those.

trevorglaser commented:
@w1ll1am23 Thanks for working on this. I imported your latest commit but as you stated without the library changes my HA is still unable to even find my Hatch. I would love to help test if there is any way to get the library changes. Thanks.

w1ll1am23 commented:
@trevorglaser sounds good, just requested @dahlb to review the library changes. Once we get those approved/uploaded to PyPI I'll update this PR to point at that new version. The changes here at the moment work for gen 2 but break gen 1 stuff so I'll have to fix that as well.

w1ll1am23 commented:
@trevorglaser okay... this code will work now @dahlb approved the library change and the new version has been bumped on PyPI

Before you test, what gen 2 product do you have? I believe the 2nd gen with a battery won't work with these changes as I think that is a different device type, but I assume it is probably using the same API as the non-battery one.

trevorglaser commented:
@w1ll1am23 I have the 2nd gen with the battery. And you are correct! I added the new library and custom component and it is still not able to connect. Although it is not throwing the same error as before so it is able to call the RestIot in the library.

w1ll1am23 commented:
Well glad we have someone with one at least should make it easier to test. I'll work on getting the changes for the + added hopefully it just works identical to the non-battery unit.

w1ll1am23 commented:
@trevorglaser let me know if you can test this https://github.com/dahlb/hatch_rest_api/pull/7

trevorglaser commented:
Here is the latest error with the latest commit.

```
Unexpected exception importing platform custom_components.ha_hatch.media_player
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/loader.py", line 722, in get_platform
    cache[full_name] = self._import_platform(platform_name)
  File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/loader.py", line 739, in _import_platform
    return importlib.import_module(f"{self.pkg_path}.{platform_name}")
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 790, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/homeassistant/.homeassistant/custom_components/ha_hatch/media_player.py", line 15, in <module>
    from .rest_media_entity import RestMediaEntity
  File "/home/homeassistant/.homeassistant/custom_components/ha_hatch/rest_media_entity.py", line 29, in <module>
    class RestMediaEntity(RestEntity, MediaPlayerEntity):
  File "/home/homeassistant/.homeassistant/custom_components/ha_hatch/rest_media_entity.py", line 34, in RestMediaEntity
    def __init__(self, rest_device: RestMini | RestPlus, config_turn_on_media):
TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

w1ll1am23 commented:
:( missed a file again. added the import to that rest_media_entity and push. Try one more time please sorry about that.

w1ll1am23 commented:
Hmm that was working for me, I'll test it out again. When you get a chance to send me logs, I'll get the battery added as well. Of you want you can PM me on the home assistant forums (w1ll1am23) with the logs, or post them here, but review them first to make sure there isn't any personal info.

w1ll1am23 commented:
Make sure you have the latest changes from this PR as well as the library PR in place. The light is working for me when using that.

trevorglaser commented:
```
2022-12-09 18:47:24.140 DEBUG (MainThread) [custom_components.ha_hatch] async setup
2022-12-09 18:47:24.214 DEBUG (MainThread) [custom_components.ha_hatch] async setup entry: {'email': ' . @ mail.com', 'password': 'FqbKKXW6'}:{}
2022-12-09 18:47:24.215 DEBUG (MainThread) [custom_components.ha_hatch] updating credentials: initial setup
2022-12-09 18:47:24.222 DEBUG (MainThread) [hatch_rest_api.util_http] sending https://data.hatchbaby.com/public/v1/login requestsending https://data.hatchbaby.com/public/v1/login request with {'email': ' . @ mail.com', 'password': '***'}
2022-12-09 18:47:24.870 DEBUG (MainThread) [hatch_rest_api.util_http] response headers:<CIMultiDict('Date': 'Sat, 10 Dec 2022 02:47:24 GMT', 'Content-Type': 'application/json; charset=UTF-8', 'Content-Length': '617', 'Connection': 'keep-alive', 'X-Hatch-Request-Id': 'U9oQrqUooG')>
2022-12-09 18:47:24.871 DEBUG (MainThread) [hatch_rest_api.util_http] response json: {'status': 'success', 'message': 'Successful login', 'payload': {'id': 3107672, 'createDate': '2022-10-29 10:46:52', 'updateDate': '2022-10-29 14:06:25', 'active': True, 'email': ' . @ mail.com', 'babies': [], 'defaultUnitOfMeasure': 'imperial', 'appType': 'iOS', 'firstName': ' ', 'lastName': ' ', 'signupSource': 'rest', 'timezone': 'America/Los_Angeles', 'timeZoneAdjust': -8, 'stage': -10000}, 'sync': 1670640439725, 'token': ‘******’}
2022-12-09 18:47:24.871 DEBUG (MainThread) [hatch_rest_api.util_http] sending https://data.hatchbaby.com/service/app/iotDevice/v2/fetch request
2022-12-09 18:47:25.315 DEBUG (MainThread) [hatch_rest_api.util_http] response headers:<CIMultiDict('Date': 'Sat, 10 Dec 2022 02:47:24 GMT', 'Content-Type': 'application/json; charset=UTF-8', 'Content-Length': '501', 'Connection': 'keep-alive', 'Expires': '0', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'X-XSS-Protection': '1; mode=block', 'Pragma': 'no-cache', 'X-Frame-Options': 'DENY', 'Etag': '"07fed7ca462d027f1e3d62c56b6e16f37"', 'X-Content-Type-Options': 'nosniff', 'X-Hatch-Request-Id': '9ypPzgjf5u')>
2022-12-09 18:47:25.315 DEBUG (MainThread) [hatch_rest_api.util_http] response json: {'status': 'success', 'message': 'Fetched list of IoT Devices', 'payload': [{'id': 1043724, 'createDate': '2022-10-29 13:59:23', 'updateDate': '2022-10-29 13:59:23', 'macAddress': ' ******', 'owner': True, 'name': ' ’s Sound Machine', 'hardwareVersion': '8.00', 'product': 'riotPlus', 'email': ' . @ mail.com', 'memberId': 3107672, 'thingName': ' ******-riotPlus'}], 'sync': 1670640439895}
2022-12-09 18:47:25.315 DEBUG (MainThread) [hatch_rest_api.util_http] sending https://data.hatchbaby.com/service/app/restPlus/token/v1/fetch request
2022-12-09 18:47:25.560 DEBUG (MainThread) [hatch_rest_api.util_http] response headers:<CIMultiDict('Date': 'Sat, 10 Dec 2022 02:47:25 GMT', 'Content-Type': 'application/json; charset=UTF-8', 'Content-Length': '1439', 'Connection': 'keep-alive', 'Expires': '0', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'X-XSS-Protection': '1; mode=block', 'Pragma': 'no-cache', 'X-Frame-Options': 'DENY', 'Etag': '"09c60b12aa9b4d44fece63ab2681c1d1f"', 'X-Content-Type-Options': 'nosniff', 'X-Hatch-Request-Id': 'R7vd4VYanB')>
2022-12-09 18:47:25.561 DEBUG (MainThread) [hatch_rest_api.util_http] response json: {'status': 'success', 'message': 'Got Cognito OpenId token', 'payload': {'endpoint': 'https://aacsjje0gbbcc-ats.iot.us-west-2.amazonaws.com', 'identityId': 'us-west-2:3c8ee252-69e7-497b-9718-7cd333d9b4ee', 'region': 'us-west-2', 'cognitoPoolId': 'us-west-2:5651fc22-f313-49d5-b379-fdab2c535d9c', 'token': ‘***’*}, 'sync': 1670640440340}
2022-12-09 18:47:25.561 DEBUG (MainThread) [hatch_rest_api.util_http] sending https://data.hatchbaby.com/service/app/routine/v2/fetch request
2022-12-09 18:47:25.813 DEBUG (MainThread) [hatch_rest_api.util_http] response headers:<CIMultiDict('Date': 'Sat, 10 Dec 2022 02:47:25 GMT', 'Content-Type': 'application/json; charset=UTF-8', 'Content-Length': '6663', 'Connection': 'keep-alive', 'Expires': '0', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'X-XSS-Protection': '1; mode=block', 'Pragma': 'no-cache', 'X-Frame-Options': 'DENY', 'Etag': '"0f9923c2df67bafc2ab762bc9993fd033"', 'X-Content-Type-Options': 'nosniff', 'X-Hatch-Request-Id': 'jShBm9vSAJ')>
2022-12-09 18:47:25.814 DEBUG (MainThread) [hatch_rest_api.util_http] response json: {'status': 'success', 'message': 'got routines', 'payload': [{'id': 15192939, 'macAddress': ' ******', 'name': 'White', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 1, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'White', 'type': 'basic', 'enabled': True, 'color': {'id': 233, 'contentfulId': '3QOi6xUa4SXs70WBRXHAK1', 'r': 0, 'g': 0, 'b': 0, 'w': 65535, 'i': 65535, 'duration': None, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 19998, 'contentfulId': '6SgX9tcTUsXBGReeE48v89', 'v': 13107, 'duration': None, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': 1}]}, {'id': 13284758, 'macAddress': ' ******', 'name': 'Chirping Birds', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 7, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 237, 'contentfulId': '1vjUCylKxo8N0uprhcUlie', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10027, 'contentfulId': '0vARayJaDqr0vTcG58IE3', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284757, 'macAddress': ' ******', 'name': 'Fan', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 6, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 236, 'contentfulId': '918IflmjYHndnIrJAmq29', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10145, 'contentfulId': '3YOWXicSuzloNsuokjiPtC', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284756, 'macAddress': ' ******', 'name': 'Wind', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 5, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 239, 'contentfulId': '6ZaxFhHc4T8YGGTlniq80Q', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10141, 'contentfulId': 'TxqKUil36cuDmqT49ZJ2X', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284755, 'macAddress': ' ******', 'name': 'Ocean', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 4, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 9998, 'contentfulId': '3A2yfBSI1hx6m4XXLjExAO', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10138, 'contentfulId': '7D76yYvto4j5FMYwpBmLkV', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284754, 'macAddress': ' ******', 'name': 'Rain', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 3, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 9997, 'contentfulId': '4SWhwGMBwFPvXgaxiIyOtf', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10139, 'contentfulId': '5hjzEGkEnsEOBtEUPosceg', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284753, 'macAddress': ' ******', 'name': 'Sleep', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 2, 'sleepScene': True, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 234, 'contentfulId': '4jVKLR0Jot7x4nJiGgO5Or', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10137, 'contentfulId': '29AT7A9MNXDnyoQI2Fv9Lk', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}], 'sync': 1670640440585}
2022-12-09 18:47:25.814 DEBUG (MainThread) [hatch_rest_api.util_bootstrap] Favorites for  ******: [{'id': 15192939, 'macAddress': ' ******', 'name': 'White', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 1, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'White', 'type': 'basic', 'enabled': True, 'color': {'id': 233, 'contentfulId': '3QOi6xUa4SXs70WBRXHAK1', 'r': 0, 'g': 0, 'b': 0, 'w': 65535, 'i': 65535, 'duration': None, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 19998, 'contentfulId': '6SgX9tcTUsXBGReeE48v89', 'v': 13107, 'duration': None, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': 1}]}, {'id': 13284758, 'macAddress': ' ******', 'name': 'Chirping Birds', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 7, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 237, 'contentfulId': '1vjUCylKxo8N0uprhcUlie', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10027, 'contentfulId': '0vARayJaDqr0vTcG58IE3', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284757, 'macAddress': ' ******', 'name': 'Fan', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 6, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 236, 'contentfulId': '918IflmjYHndnIrJAmq29', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10145, 'contentfulId': '3YOWXicSuzloNsuokjiPtC', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284756, 'macAddress': ' ******', 'name': 'Wind', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 5, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 239, 'contentfulId': '6ZaxFhHc4T8YGGTlniq80Q', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10141, 'contentfulId': 'TxqKUil36cuDmqT49ZJ2X', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284755, 'macAddress': ' ******', 'name': 'Ocean', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 4, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 9998, 'contentfulId': '3A2yfBSI1hx6m4XXLjExAO', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10138, 'contentfulId': '7D76yYvto4j5FMYwpBmLkV', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284754, 'macAddress': ' ******', 'name': 'Rain', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 3, 'sleepScene': False, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 9997, 'contentfulId': '4SWhwGMBwFPvXgaxiIyOtf', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10139, 'contentfulId': '5hjzEGkEnsEOBtEUPosceg', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}, {'id': 13284753, 'macAddress': ' ******', 'name': 'Sleep', 'type': 'favorite', 'active': True, 'enabled': True, 'displayOrder': 2, 'sleepScene': True, 'followBySleepScene': False, 'startTime': None, 'endTime': None, 'daysOfWeek': 0, 'steps': [{'name': 'Step 1', 'type': 'basic', 'enabled': True, 'color': {'id': 234, 'contentfulId': '4jVKLR0Jot7x4nJiGgO5Or', 'r': None, 'g': None, 'b': None, 'w': None, 'i': 26000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'sound': {'id': 10137, 'contentfulId': '29AT7A9MNXDnyoQI2Fv9Lk', 'v': 13000, 'duration': 0, 'until': 'indefinite', 'ignore': False}, 'mixId': None, 'dayOffset': None}]}]
2022-12-09 18:47:25.814 DEBUG (MainThread) [hatch_rest_api.util_http] sending https://cognito-identity.us-west-2.amazonaws.com requestheaders: {'content-type': 'application/x-amz-json-1.1', 'X-Amz-Target': 'AWSCognitoIdentityService.GetCredentialsForIdentity'}sending https://cognito-identity.us-west-2.amazonaws.com request with {'IdentityId': 'us-west-2:3c8ee252-69e7-497b-9718-7cd333d9b4ee', 'Logins': {'cognito-identity.amazonaws.com': ‘***’*}}
2022-12-09 18:47:26.801 DEBUG (MainThread) [hatch_rest_api.util_http] response headers:<CIMultiDict('Date': 'Sat, 10 Dec 2022 02:47:26 GMT', 'Content-Type': 'application/x-amz-json-1.1', 'Content-Length': '1404', 'Connection': 'keep-alive', 'x-amzn-RequestId': '4aa4f005-0cda-4d1c-93a1-9bf0806ff74a', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains')>
2022-12-09 18:47:26.815 DEBUG (MainThread) [hatch_rest_api.util_http] response raw: {"Credentials":{"AccessKeyId":"*****","Expiration":*****,"SecretKey”:”**”**},”IdentityId”:”**”***}
2022-12-09 18:47:27.228 DEBUG (MainThread) [hatch_rest_api.util_bootstrap] mqtt connection connected
2022-12-09 18:47:27.228 DEBUG (MainThread) [hatch_rest_api.shadow_client_subscriber] creating RestIot:  ’s Sound Machine
2022-12-09 18:47:27.313 DEBUG (MainThread) [hatch_rest_api.shadow_client_subscriber] unsubscribe_topic_to_update_shadow_accepted: $aws/things/ ******-riotPlus/shadow/update/accepted
2022-12-09 18:47:27.432 DEBUG (MainThread) [hatch_rest_api.shadow_client_subscriber] unsubscribe_topic_to_update_shadow_accepted: $aws/things/ ******-riotPlus/shadow/get/accepted
2022-12-09 18:47:27.432 DEBUG (MainThread) [hatch_rest_api.shadow_client_subscriber] Requesting current shadow state...
2022-12-09 18:47:27.475 DEBUG (MainThread) [hatch_rest_api.shadow_client_subscriber] result: None
2022-12-09 18:47:27.475 DEBUG (MainThread) [custom_components.ha_hatch] credentials expire at: 2022-12-09 19:47:26
2022-12-09 18:47:27.518 DEBUG (Dummy-2) [hatch_rest_api.shadow_client_subscriber] get  ’s Sound Machine, RESPONSE: awsiot.iotshadow.GetShadowResponse(client_token=None, metadata=awsiot.iotshadow.ShadowMetadata(desired={'owned': {'timestamp': 1667077162}, 'timezone': {'timestamp': 1670381368}, 'env': {'timestamp': 1667077162}, 'logging': {'timestamp': 1669726976}, 'clock': {'flags': {'timestamp': 1670381368}, 'i': {'timestamp': 1670381368}, 'turnOffMode': {'timestamp': 1670381368}, 'turnOffAt': {'timestamp': 1670381368}, 'turnOnAt': {'timestamp': 1670381368}}, 'current': {'color': {'i': {'timestamp': 1670639228}, 'id': {'timestamp': 1670639228}, 'r': {'timestamp': 1670639228}, 'g': {'timestamp': 1670639228}, 'b': {'timestamp': 1670639228}, 'w': {'timestamp': 1670639228}, 'duration': {'timestamp': 1670639214}, 'until': {'timestamp': 1670639214}}, 'sound': {'v': {'timestamp': 1670639214}, 'id': {'timestamp': 1670639214}, 'mute': {'timestamp': 1670639214}, 'url': {'timestamp': 1670639214}, 'duration': {'timestamp': 1670639214}, 'until': {'timestamp': 1670639214}}, 'srId': {'timestamp': 1670639217}, 'playing': {'timestamp': 1670639217}, 'step': {'timestamp': 1670639217}}, 'alarmsDisabled': {'timestamp': 1670381368}, 'nightlightOn': {'timestamp': 1670381368}, 'snoozeDuration': {'timestamp': 1670381368}, 'sleepScene': {'enabled': {'timestamp': 1670639214}}, 'timer': {'s': {'timestamp': 1670639214}, 'd': {'timestamp': 1670639214}}, 'rF': {'i': {'timestamp': 1670381368}, 'u': {'timestamp': 1670381368}, 'v': {'timestamp': 1670381368}}, 'LWTP': {'timestamp': 1668896569}, 'toddlerLock': {'turnOffAt': {'timestamp': 1670381368}, 'turnOnAt': {'timestamp': 1670381368}, 'turnOnMode': {'timestamp': 1670381368}}, 'lucky': {'timestamp': 1670383026}}, reported={'env': {'timestamp': 1670639214}, 'alarmsDisabled': {'timestamp': 1670633105}, 'nightlightOn': {'timestamp': 1670633105}, 'nightlightIntensity': {'timestamp': 1670633105}, 'toddlerLockOn': {'timestamp': 1670633105}, 'snoozeDuration': {'timestamp': 1670633105}, 'current': {'srId': {'timestamp': 1670639214}, 'playing': {'timestamp': 1670639218}, 'step': {'timestamp': 1670639214}, 'color': {'id': {'timestamp': 1670639229}, 'r': {'timestamp': 1670639229}, 'g': {'timestamp': 1670639229}, 'b': {'timestamp': 1670639229}, 'w': {'timestamp': 1670633105}, 'i': {'timestamp': 1670639229}, 'duration': {'timestamp': 1670633105}, 'until': {'timestamp': 1670633105}}, 'sound': {'id': {'timestamp': 1670639214}, 'v': {'timestamp': 1670639208}, 'mute': {'timestamp': 1670633105}, 'url': {'timestamp': 1670639214}, 'duration': {'timestamp': 1670633105}, 'until': {'timestamp': 1670633105}}}, 'dataVersion': {'timestamp': 1670633105}, 'sleepScene': {'srId': {'timestamp': 1670633105}, 'enabled': {'timestamp': 1670633105}}, 'timer': {'s': {'timestamp': 1670633105}, 'd': {'timestamp': 1670633105}}, 'timezone': {'timestamp': 1670633105}, 'rF': {'v': {'timestamp': 1670633105}, 'i': {'timestamp': 1670633105}, 'u': {'timestamp': 1670633105}}, 'deviceInfo': {'b': {'timestamp': 1670633105}, 'f': {'timestamp': 1670633105}, 'fR': {'timestamp': 1670633105}, 'hwVersion': {'timestamp': 1670633105}, 'powerStatus': {'timestamp': 1668835185}}, 'clock': {'i': {'timestamp': 1670639214}, 'turnOffAt': {'timestamp': 1670639214}, 'turnOnAt': {'timestamp': 1670639214}, 'flags': {'timestamp': 1670639214}, 'turnOffMode': {'timestamp': 1670639214}}, 'toddlerLock': {'turnOffAt': {'timestamp': 1670639214}, 'turnOnAt': {'timestamp': 1670639214}, 'turnOnMode': {'timestamp': 1670639214}}, 'lucky': {'timestamp': 1670633105}, 'LDR': {'timestamp': 1670633105}, 'LWTP': {'timestamp': 1670639214}, 'debug': {'timestamp': 1667077365}, 'logging': {'timestamp': 1669726977}, 'owned': {'timestamp': 1670633105}, 'lastReset': {'timestamp': 1670633105}, 'REX': {'lock': {'timestamp': 1670633105}, 'key': {'timestamp': 1670633105}, 'command': {'timestamp': 1670633105}}, 'connected': {'timestamp': 1670639229}, 'rssi': {'timestamp': 1670639229}, 'streaming': {'status': {'timestamp': 1670633105}}, 'ota': {'status': {'timestamp': 1670633105}, 'downloadProgress': {'timestamp': 1670633105}, 'installProgress': {'timestamp': 1670633105}}}), state=awsiot.iotshadow.ShadowStateWithDelta(delta=None, desired={'owned': True, 'timezone': 'America/Los_Angeles', 'env': 'prod', 'logging': 21, 'clock': {'flags': 32768, 'i': 655, 'turnOffMode': 'never', 'turnOffAt': '22:00:00', 'turnOnAt': '07:00:00'}, 'current': {'color': {'i': 0, 'id': 9998, 'r': 0, 'g': 0, 'b': 0, 'w': 0, 'duration': 0, 'until': 'indefinite'}, 'sound': {'v': 13000, 'id': 19998, 'mute': False, 'url': 'https://images.hatchbaby.com/content/sounds/noSound.wav', 'duration': 0, 'until': 'indefinite'}, 'srId': 0, 'playing': 'remote', 'step': 0}, 'alarmsDisabled': False, 'nightlightOn': False, 'snoozeDuration': 540, 'sleepScene': {'enabled': False}, 'timer': {'s': '2022-12-06 18:40:47', 'd': 0}, 'rF': {'i': True, 'u': 'https://firmware.hatchbaby.com/prod/riotPlus/riotPlus_DFU_7_1_368.bin?f=g', 'v': '7.1.368'}, 'LWTP': False, 'toddlerLock': {'turnOffAt': '07:00:00', 'turnOnAt': '19:00:00', 'turnOnMode': 'never'}, 'lucky': 31432}, reported={'env': 'prod', 'alarmsDisabled': False, 'nightlightOn': False, 'nightlightIntensity': 32780, 'toddlerLockOn': False, 'snoozeDuration': 540, 'current': {'srId': 0, 'playing': 'remote', 'step': 0, 'color': {'id': 9998, 'r': 0, 'g': 0, 'b': 0, 'w': 0, 'i': 0, 'duration': 0, 'until': 'indefinite'}, 'sound': {'id': 19998, 'v': 13000, 'mute': False, 'url': 'https://images.hatchbaby.com/content/sounds/noSound.wav', 'duration': 0, 'until': 'indefinite'}}, 'dataVersion': '20221205173803704', 'sleepScene': {'srId': 0, 'enabled': False}, 'timer': {'s': '2022-12-06 18:40:47', 'd': 0}, 'timezone': 'America/Los_Angeles', 'rF': {'v': '7.1.368', 'i': True, 'u': 'https://firmware.hatchbaby.com/prod/riotPlus/riotPlus_DFU_7_1_368.bin?f=g'}, 'deviceInfo': {'b': 100, 'f': '7.1.368', 'fR': 0, 'hwVersion': '8.0.0', 'powerStatus': 5}, 'clock': {'i': 655, 'turnOffAt': '22:00:00', 'turnOnAt': '07:00:00', 'flags': 32768, 'turnOffMode': 'never'}, 'toddlerLock': {'turnOffAt': '07:00:00', 'turnOnAt': '19:00:00', 'turnOnMode': 'never'}, 'lucky': 31432, 'LDR': 'OK', 'LWTP': False, 'debug': 0, 'logging': 21, 'owned': True, 'lastReset': 'DeepSleep', 'REX': {'lock': 655938258, 'key': 0, 'command': 'none'}, 'connected': True, 'rssi': -58, 'streaming': {'status': 'none'}, 'ota': {'status': 'none', 'downloadProgress': 0, 'installProgress': 0}}), timestamp=datetime.datetime(2022, 12, 9, 18, 47, 27), version=563)
2022-12-09 18:47:27.518 DEBUG (Dummy-2) [hatch_rest_api.riot] update local state:  ’s Sound Machine, {'env': 'prod', 'alarmsDisabled': False, 'nightlightOn': False, 'nightlightIntensity': 32780, 'toddlerLockOn': False, 'snoozeDuration': 540, 'current': {'srId': 0, 'playing': 'remote', 'step': 0, 'color': {'id': 9998, 'r': 0, 'g': 0, 'b': 0, 'w': 0, 'i': 0, 'duration': 0, 'until': 'indefinite'}, 'sound': {'id': 19998, 'v': 13000, 'mute': False, 'url': 'https://images.hatchbaby.com/content/sounds/noSound.wav', 'duration': 0, 'until': 'indefinite'}}, 'dataVersion': '20221205173803704', 'sleepScene': {'srId': 0, 'enabled': False}, 'timer': {'s': '2022-12-06 18:40:47', 'd': 0}, 'timezone': 'America/Los_Angeles', 'rF': {'v': '7.1.368', 'i': True, 'u': 'https://firmware.hatchbaby.com/prod/riotPlus/riotPlus_DFU_7_1_368.bin?f=g'}, 'deviceInfo': {'b': 100, 'f': '7.1.368', 'fR': 0, 'hwVersion': '8.0.0', 'powerStatus': 5}, 'clock': {'i': 655, 'turnOffAt': '22:00:00', 'turnOnAt': '07:00:00', 'flags': 32768, 'turnOffMode': 'never'}, 'toddlerLock': {'turnOffAt': '07:00:00', 'turnOnAt': '19:00:00', 'turnOnMode': 'never'}, 'lucky': 31432, 'LDR': 'OK', 'LWTP': False, 'debug': 0, 'logging': 21, 'owned': True, 'lastReset': 'DeepSleep', 'REX': {'lock': 655938258, 'key': 0, 'command': 'none'}, 'connected': True, 'rssi': -58, 'streaming': {'status': 'none'}, 'ota': {'status': 'none', 'downloadProgress': 0, 'installProgress': 0}}
2022-12-09 18:47:27.518 DEBUG (Dummy-2) [hatch_rest_api.riot] new state:{'device_name': ' ’s Sound Machine', 'thing_name': ' ******-riotPlus', 'firmware_version': '7.1.368', 'is_online': True, 'is_on': False, 'is_playing': False, 'volume': 19, 'red': 0, 'green': 0, 'blue': 0, 'brightness': 0, 'document_version': 563}
2022-12-09 18:47:27.518 DEBUG (Dummy-2) [hatch_rest_api.callbacks]  ’s Sound Machine publishing updates
2022-12-09 18:47:27.553 DEBUG (MainThread) [custom_components.ha_hatch.light] setting up hatch lights, auto turn on switch set to True
2022-12-09 18:47:27.734 DEBUG (MainThread) [custom_components.ha_hatch.riot_media_entity] updating state:{'device_name': ' ’s Sound Machine', 'thing_name': ' ******-riotPlus', 'firmware_version': '7.1.368', 'is_online': True, 'is_on': False, 'is_playing': False, 'volume': 19, 'red': 0, 'green': 0, 'blue': 0, 'brightness': 0, 'document_version': 563}
2022-12-09 18:47:27.735 DEBUG (MainThread) [custom_components.ha_hatch.binary_sensor] updating state:{'device_name': ' ’s Sound Machine', 'thing_name': ' ******-riotPlus', 'firmware_version': '7.1.368', 'is_online': True, 'is_on': False, 'is_playing': False, 'volume': 19, 'red': 0, 'green': 0, 'blue': 0, 'brightness': 0, 'document_version': 563}
2022-12-09 18:47:27.758 DEBUG (MainThread) [custom_components.ha_hatch.light] updating state:{'device_name': ' ’s Sound Machine', 'thing_name': ' ******-riotPlus', 'firmware_version': '7.1.368', 'is_online': True, 'is_on': False, 'is_playing': False, 'volume': 19, 'red': 0, 'green': 0, 'blue': 0, 'brightness': 0, 'document_version': 563}
```

trevorglaser commented:
@w1ll1am23 the b is battery. and powerStatus 5 is "Charging, on base"

` 'deviceInfo': {'b': 100, 'f': '7.1.368', 'fR': 0, 'hwVersion': '8.0.0', 'powerStatus': 5},`

w1ll1am23 commented:
Nice! I'll work on adding this in, hopefully some time tomorrow.

w1ll1am23 commented:
@trevorglaser decided to do it now, might not have much time tomorrow. Push new changes to both PRs.

Just going to keep the battery consistent with the gen 1 and only report the percentage. Once we have the main functions working I'll look in to reporting the other data, like the night light, clock, and the charging status.

trevorglaser commented:
@w1ll1am23 I am not seeing the battery percentage reported as a sensor in HA.

w1ll1am23 commented:
Pushed an updated sensors.py I added a check to not add a battery sensor for the non-battery versions but I assume that prevented it from showing up for you... going to have to figure that one out.

just double checking you put in both changes the library and this customer component code?

trevorglaser commented:
@w1ll1am23 you forgot to check back in sensor.py in the latest commit. Yes I did both the library and custom component.

w1ll1am23 commented:
I have no idea how I managed to do that lol

trevorglaser commented:
Still not showing the sensor.

```
2022-12-09 21:47:47.585 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up ha_hatch platform for sensor
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/helpers/entity_platform.py", line 281, in _async_setup_platform
    await asyncio.shield(task)
  File "/home/homeassistant/.homeassistant/custom_components/ha_hatch/sensor.py", line 23, in async_setup_entry
    if isinstance(rest_device, RestPlus | RestIot ):
TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

w1ll1am23 commented:
@trevorglaser pull the sensor.py one more time, this should hopefully work.

At some point I'll have to think on this some more and figure out how to now show the battery on the non-battery devices.

trevorglaser commented:
Its working and displaying the battery now.

w1ll1am23 commented:
@dahlb this is now ready for review as well. Would also be cool if you could double check that the v1 stuff is still working since @trevorglaser and I don't have one of those to test.

dahlb commented on custom_components/ha_hatch/light.py:
```diff
@@ -41,15 +43,18 @@ class HatchLightEntity(RestEntity, LightEntity):
         if self.platform is None:
             return
         _LOGGER.debug(f"updating state:{self.rest_device}")
-        self._attr_is_on = self.rest_device.is_on
```
this looks like an unintentional change to the old behavior, the original code should be in this else but instead is in the if

w1ll1am23 replied:
Moved the new gen 2 light to its own object so this no longer applies.

dahlb commented on custom_components/ha_hatch/light.py:
```diff
@@ -61,17 +66,27 @@ def turn_on(self, **kwargs):
         else:
-            brightness = self._attr_brightness
+            if isinstance(self.rest_device, RestIot):
+                brightness = 0
```
why does this default to turning on at 0 brightness instead of the old known brightness when brightness is not specified?

w1ll1am23 replied:
I can probably make this more clear with the changes you mentioned below, having the previous brightness here will allow us to pass that.

dahlb commented on custom_components/ha_hatch/light.py:
```diff
@@ -61,17 +66,27 @@ def turn_on(self, **kwargs):
         _LOGGER.debug(f"turning on light to {rgb} with {brightness}")
-        self.rest_device.set_color(rgb[0], rgb[1], rgb[2], brightness)
+        if isinstance(self.rest_device, RestIot):
+            self.rest_device.set_color(rgb[0], rgb[1], rgb[2], 0, brightness, True)
```
I don't think this logic will hold up, the rest_devices are thrown away when updating expired auth so any state you want to preserve should be handled in this repo as the api objects are not persistent

w1ll1am23 replied:
Hmm I didn't realize this. So you are saying the cached light values in the library won't exist anymore because that object gets replaced? Where is that expired auth logic at? How is HA made aware of the new object?

dahlb replied:
RestEntity, replace_rest_device could be overridden here if you want to preserve it during replacement. it isn't the easiest code to follow but is invoked in __init__.py setup_connection method

w1ll1am23 replied:
I see, okay I'll work on this.

w1ll1am23 replied:
I believe this should be more clear now having split it in to multiple objects

dahlb commented on custom_components/ha_hatch/light.py:
```diff
@@ -61,17 +66,27 @@ def turn_on(self, **kwargs):
-        self.rest_device.set_color(self.rest_device.red, self.rest_device.green, self.rest_device.blue, 0)
+        if isinstance(self.rest_device, RestIot):
+            _LOGGER.debug("turning off with rgbw = 0 and bright = 0")
+            self.rest_device.set_color(0, 0, 0, 0, 0, False)
```
looks like you are having to modify the logic in all of the methods of HatchLightEntity, would be easier to maintain if we just make two classes, maybe HatchLightPlusEntity and HatchLightIotEntity

dahlb replied:
it seems confusing to have 5 parameters that will never be needed if the 6th is, did you consider separating the set_color from turn_off_light instead? it doesn't match the plus either way because i didn't expose the w value and am curious how 0 brightness works differently in the iot device from the custom color id, looks like the gen2 api makes it harder to specify the light attributes without also having to set the sound (saw that weird remote thing too) I haven't worked with that api so I don't feel strongly either way.

w1ll1am23 replied:
Yeah, this is a good idea.

w1ll1am23 replied:
> looks like you are having to modify the logic in all of the methods of HatchLightEntity, would be easier to maintain if we just make two classes, maybe HatchLightPlusEntity and HatchLightIotEntity

Yeah this is a good idea.

w1ll1am23 replied:
> it seems confusing to have 5 parameters that will never be needed if the 6th is, did you consider separating the set_color from turn_off_light instead? it doesn't match the plus either way because i didn't expose the w value and am curious how 0 brightness works differently in the iot device from the custom color id, looks like the gen2 api makes it harder to specify the light attributes without also having to set the sound (saw that weird remote thing too) I haven't worked with that api so I don't feel strongly either way.

Yeah they don't make it easy to change just one thing. Should be able to clean this up with the new entity like you suggested.

w1ll1am23 replied:
Split the on/color change and turning off functions in the library so it should make this logic more clear.

dahlb commented on custom_components/ha_hatch/riot_media_entity.py:
```diff
@@ -0,0 +1,69 @@
+        self.rest_device.set_favorite(sound_mode)
+
+    def media_stop(self):
+        self.rest_device.turn_off()
```
when this media stops does it turn off the light and sound or just the sound? if it is just the sound the turn_off might be better named media_stop, if it is both you might want to update the state of the light entity too, unless it's already doing that and I missed it

w1ll1am23 replied:
Unfortunately media stop will basically turn off everything. There is a way to just turn off the media by setting the sound to the no sound wav file, but there is no way of knowing what that URL is without calling another API they are using https://www.contentful.com/ and trying to get that working seem like a lot of work/maybe not possible. 

The light state does get updated because of the published state coming back from the API

w1ll1am23 commented:
@dahlb this change along with https://github.com/dahlb/hatch_rest_api/pull/9 should be good for another review now. 

Basically did what you suggested and split the rest and riot lights in to their own objects vs doing a bunch of type checks, and moved the previous light statue cache over here from the library to deal with the rest_devices getting swapped out when the auth expires.

trevorglaser commented:
@w1ll1am23 When you bricked your hatch how did you get it working again? I sent bad data to the hatch trying to add more functionality and now it wont connect anymore.

w1ll1am23 commented:
@trevorglaser you didn't have that issue using any of this code here right?

Unfortunately the only way I found to get it fixed was to email support and have them remove the device from my account. You can't reset it yourself anywhere in the app, and you can't pair it to a new account (very annoying) I didn't try deleting my account entirely that might work. 

They required a picture of the bottom of the device and proof of purchase and of course made me try all sorts of things over two days before they would reset the device.

dahlb commented:
looks good, but you'll need to update the api version to 1.17.0 in __init__.py before we merge this

trevorglaser commented:
@w1ll1am23 No it wasn't with the code here. I took the code and added a toddler lock switch, the clock light, and charge status to it. I planned on creating a PR once your code was finalized.

Thank you. I have gone back and forth for days but they haven't given me any solutions. I will ask them to remove it from my account.

w1ll1am23 commented:
@dahlb version bumped did one last smoke test and found a small issue with the brightness that I think probably impacts gen 1 if you can confirm I can fix that here as well.

w1ll1am23 commented:
> @w1ll1am23 No it wasn't with the code here. I took the code and added a toddler lock switch, the clock light, and charge status to it. I planned on creating a PR once your code was finalized.
> 
> Thank you. I have gone back and forth for days but they haven't given me any solutions. I will ask them to remove it from my account.

Ugh that is annoying I bricked mine twice and was worried they wouldn't fix it the second time. I was more cautious after that and validated everything the official Android app was sending before passing things. If they just had a way to remove the device from your account yourself it wouldn't be that big of a deal. I think the issue is the bad data is getting retained on the MQTT server and the topic has to be removed to wipe that out.
